madmom/features/beats.py

```diff
@@ -1238,6 +1238,380 @@ def add_arguments(parser, min_bpm=MIN_BPM, max_bpm=MAX_BPM,
+        of all previously detected beats.
+
+        Parameters
+        ----------
+        inner_window : int
+            Inner window in frames to each side where a beat is accepted.
+        outer_window : float
+            outer window factor to each side depending on the interval.
+        threshold : float
+            Threshold value for accepting a beat.
+        correction_factor : float
+            Allow agent to adapt to errors.
+        inherit_score_factor : float
+            Child agents will inherit a percentage of their parents.
+
+        """
+        # TODO: Should we use @classmethod to set all values from outside?
+        INNER_WINDOW = 5
+        OUTER_WINDOW = 0.4
+        THRESHOLD = 0.05
+        CORRECTION_FACTOR = 0.25
+        INHERIT_SCORE_FACTOR = 0.9
+
+        # used for normalizing the score, set from outside
+        _MAX_INTERVAL = None
+        _MIN_INTERVAL = None
+
+        def __init__(self, score=0, interval=0, prediction=0, beats=[]):
+            self.score = score
+            self.interval = interval
+            self.prediction = prediction
+            self.beats = beats
+
+        def __hash__(self):
+            return hash(self.prediction + self.interval)
+
+        def __eq__(self, other):
+            # TODO: Maybe allow for slight variations here
+            eq_prediction = self.prediction == other.prediction
+            eq_interval = self.interval == other.interval
+            return eq_prediction and eq_interval
+
+        def fork(self, error):
+            """
+            Return child agents based on the given error. The childs
+            inherit a part of the parents score and all of the parents
+            detections. The interval and the prediction for the next beat
+            are adjusted to the error of the last prediction.
+
+            """
+            # TODO: Check whether all childs are really needed
+            agents = [
+                # create agent with same tempo but adjusted prediction
+                self.__class__(score=self.score * self.INHERIT_SCORE_FACTOR,
+                               interval=self.interval,
+                               prediction=self.prediction + error,
+                               beats=self.beats),
+                # create agent with adjusted tempo and prediction
+                self.__class__(score=self.score * self.INHERIT_SCORE_FACTOR,
+                               interval=self.interval + error,
+                               prediction=self.prediction + error,
+                               beats=self.beats),
+                # create agent with adjusted tempo and prediction by half
+                self.__class__(score=self.score * self.INHERIT_SCORE_FACTOR,
+                               interval=self.interval + int(error * 0.5),
+                               prediction=self.prediction + int(error * 0.5),
+                               beats=self.beats)
+            ]
```

Comment:
This could be done in less lines of code with a simple loop.

Maybe a stupid question, but why is the error always added and not subtracted as well?

Reply:
Hmm, since all three objects created here have different intervals and predictions the only loop-thingy coming to my mind right now would be something like that:

```
predictions = [self.prediction + error, self.prediction + error, self.prediction + int(error * 0.5)]
intervals = [self.interval, self.interval + error, self.interval + int(error * 0.5)]

for p, i in zip(predictions, intervals):
  self.__class__(prediction=p, interval=i...)
```
Is that would you mean? I also want to test if all three child-agents are really needed (they are used in the paper) to make the results better.

Regarding the error variable: It can also be a negative number (if the event occurred before the prediction), so adding should work ;)

Reply:
Yes, something like this, but you could even simplify it to:

```
prediction_errors = [1, 1, 0.5]
interval_errors = [0, 1, 0.5]

for p, i in zip(prediction_errors, interval_errors):
    self.__class__(prediction=self.prediction + int(p * error),
                   interval=self.interval + int(i * error), ...)
```

Reply:
Nice, thx!